### PR TITLE
Add Fastlane Actions for Release Checks Migration

### PR DESCRIPTION
**Pull request by ravenstewart**
This pull request makes progress towards: [WordPress-Android issue #10699 (Migrate Release Checks to Ruby)](https://github.com/wordpress-mobile/WordPress-Android/issues/10699)

### Description
I added `fastlane` actions to help with migration of [`release-checks.sh`](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/tools/release-checks.sh).

The script we want to migrate has the following functionality:
- verifies that any languages that have reached >90% translation are included in the project
- verifies that `strings.xml` contains all of the strings from the project
- prints the current version from `build.gradle` (not currently working properly)
- checks that there are devices connected then runs the test suite (not currently used)

I added the following `fastlane` actions:
- **android_check_devices**
  - **parameters:** 
    - `require_device `- whether devices are required
  - **actions:**
    - prints out a list of devices, fails the check if devices are required but no devices are present
- **android_check_en_strings**
  - **parameters:**
    - `res_dir `- the res directory containing `values/strings.xml`
  - **actions:**
    - ensures clean git status
    - temporarily deletes all localized strings.xml
    - makes a test build to verify that no strings are missing
    - cleans up after test build completes
    - displays test results, fails the check if the build fails
  - **notes:**
    - If a build fails, this action currently displays the build errors so that the user knows what failed. This is different from the original script which does not display any output from the build. If this is not preferred, please let me know and I am happy to display the missing strings a different way.
    - I decided to require a clean git status rather than stashing and restoring local changes.  This behavior differs from the original script.  If this is not preferred, please let me know and I am happy to model the original behavior.
- **android_check_new_languages_glotpress**
  - **parameters:**
    - `glotpress_status_url` - url with GlotPress language translation status
    - `language_file` - language file specifying languages included in the project
  - **actions:**
    - grabs languages with > 90% status from GlotPress language translation status web page
    - checks the language file for each language, prints out whether the language was found
    - displays test results and fails the check if a language is > 90% translated and not in the language file
- **android_run_gradlew**
  - **parameters:**
    - `command` - the command to run gradlew with
  - **actions:**
    - runs ./gradlew with the specified command
    - displays results and the action fails if the build fails

I also added a couple of utility functions to [`android_localize_helper`](https://github.com/ravenstewart/release-toolkit/blob/develop/lib/fastlane/plugin/wpmreleasetoolkit/helper/android_localize_helper.rb):
- `get_glotpress_languages_translated_morethan90`
- `get_missing_languages`

After looking at it again while writing this PR, I am unsure whether these are too specific to belong in this helper...

### Testing
I tested these using the following steps:
- copy the modified files manually into the installed plug-in folder in my local `WordPress-Android` project
- use `bundle exec fastlane run <action> ...`

### Next Steps
- figure out which version # needs to be displayed
  - `release-checks.sh` no longer prints out the correct version due to some changes made to `WordPress/build.gradle`. I believe when the `printVersion` function was first written, `build.gradle` only had one `versionName` specified. Now there is one for `defaultConfig` and one for the `vanilla` flavor.  I'm not sure which we should display for the release check.
- create an action that uses functions in the existing [`android_version_helper`](https://github.com/wordpress-mobile/release-toolkit/blob/develop/lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_version_helper.rb) to print out the correct version
- use the newly created actions in the `WordPress-Android` project and replace `release-checks.sh`

### CODE REVIEW

**Comment by ravenstewart** [comment]
Hi @jkmassel !

I made the following changes to address items from the feedback:
* removed all uses of the `sh` method (please let me know if I missed anything)
* added test coverage for new functionality
  * I believe I have addressed all functionality I added, but please let me know if that is not the case!
* removed redundant `gradle` action I added

Next I will be working on using the actions from this PR in a lane in the `WordPress-Android` project!

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_devices.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
+module Fastlane
```
Comment (jkmassel) [inline lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_devices.rb]: This looks like a solid action, but I can't see it being called from anywhere 🙈 

Am I just missing something? 🙃

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_update_metadata.rb [hunk 2]
```diff
@@ -9,7 +9,6 @@ def self.run(params)
         require_relative '../../helper/android/android_git_helper.rb'
 
         Fastlane::Helpers::AndroidGitHelper.update_metadata(ENV["validate_translations"])
-        Fastlane::Action::sh("./tools/release-checks.sh")
```
[the remark this answers is not in the document]
Comment (ravenstewart) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_update_metadata.rb]: The release checks actions were added to the `WordPress-Android` project in this [pull request](https://github.com/wordpress-mobile/WordPress-Android/pull/10841) 🙂

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/android_localize_helper.rb [hunk 3]
```diff
@@ -47,7 +47,7 @@ def self.merge_string(main_strings, library, string_line)
             return :skipped if skip_string_by_tag(this_string)
             
             # Update if needed
-            (if (this_string.content == string_content) then return :found else this_string.content = string_content ; return :updated end)   
+            (if (this_string.content == string_content) then return :found else this_string.content = string_content ; return :updated end)
```
[the remark this answers is not in the document]
Comment (ravenstewart) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/android_localize_helper.rb]: I will undo that! Thank you!

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/android_localize_helper.rb [hunk 4]
```diff
@@ -68,7 +68,7 @@ def self.verifystring(main_strings, library, string_line)
         return if skip_string_by_exclusion_list(library, string_name)
       
         # Search for the string in the main file
-        main_strings.xpath('//string').each do | this_string | 
+        main_strings.xpath('//string').each do | this_string |
```
[the remark this answers is not in the document]
Comment (ravenstewart) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/android_localize_helper.rb]: I will undo that! Thank you!

File: spec/test-data/malformed.csv [hunk 5]
```diff
@@ -0,0 +1,8 @@
+¢£¤¥¦§¨©ª«¬®¯ °±²ɇɈɉɊɋɌɎɏɐɑɒɓɔ ɕɖɗɘəɚ⤚▓⤜⤝⤞⤟ⰙⰚⰛⰜ⭑⬤⭒‰ ꕢ ꕣꕤ ꕥ￥￦
+        ❌ ⛱⛲⛳⛰⛴⛵ ⚡⏰⏱⏲⭐ ✋☕⛩⛺⛪✨ ⚽ ⛄⏳
```
[the remark this answers is not in the document]
Comment (ravenstewart) [reply spec/test-data/malformed.csv]: I will convert all the test data to test files.  Thank you!

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_devices.rb (hunk 1 above)
Comment (ravenstewart) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_devices.rb]: 🙂 This action is used in this [pull request](https://github.com/wordpress-mobile/WordPress-Android/pull/10841).

File: Gemfile.lock [hunk 6]
```diff
@@ -1,7 +1,7 @@
 PATH
   remote: .
   specs:
-    fastlane-plugin-wpmreleasetoolkit (0.8.0)
+    fastlane-plugin-wpmreleasetoolkit (0.8.1)
```
[the remark this answers is not in the document]
Comment (ravenstewart) [reply Gemfile.lock]: I will undo this!  Thank you!

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_en_strings.rb [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+          if params[:ensure_git_status_clean] then
+            other_action.ensure_git_status_clean(show_uncommitted_changes:true)
+          end
+
+          # set up test by removing localized strings.xml
+          deleted_files = Fastlane::Helper::FilesystemHelper::delete_files("#{params[:res_dir]}/values-??/strings.xml", true, false)
+          deleted_files += Fastlane::Helper::FilesystemHelper::delete_files("#{params[:res_dir]}/values-??-r??/strings.xml", true, false)
+
+          if (deleted_files.count > 0)
+            plural = ""
```
[the remark this answers is not in the document]
Comment (ravenstewart) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_en_strings.rb]: I will look into this!  Thank you!!

File: Gemfile.lock (hunk 6 above)
[the remark this answers is not in the document]
Comment (ravenstewart) [reply Gemfile.lock]: removed these changes here: da4ea63588b5b3812d6c881b32fd96949180339a

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/android_localize_helper.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (ravenstewart) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/android_localize_helper.rb]: removed here: f76ee6e8b6306297e722f112ae583dc991f34b2d

File: lib/fastlane/plugin/wpmreleasetoolkit/helper/android_localize_helper.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (ravenstewart) [reply lib/fastlane/plugin/wpmreleasetoolkit/helper/android_localize_helper.rb]: removed here: f76ee6e8b6306297e722f112ae583dc991f34b2d

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_en_strings.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (ravenstewart) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_en_strings.rb]: Now using pluralize (e121db9e760147acf3c67cadf00870b34d31baae)!  Thank you for the suggestion. :)

File: lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_en_strings.rb [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+          if (not success)
+            UI.user_error!("Check Failed: some English strings may be missing in #{params[:res_dir]}/values/strings.xml")
+          end
+
+          UI.success "Check Success: no missing English strings found in #{params[:res_dir]}/values/strings.xml"
+          "Check Success"
+        end
+
+        def self.run_gradle(gradlew_path, project_dir, task, print_output)
```
[the remark this answers is not in the document]
Comment (ravenstewart) [reply lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_check_en_strings.rb]: @jkmassel I decided to keep this function after removing some parameters here: a13710b9489e0a5f4fad406b60f971dc3702140a.  I think the function now provides some value, but please let me know what you think!

File: spec/test-data/malformed.csv (hunk 5 above)
[the remark this answers is not in the document]
Comment (ravenstewart) [reply spec/test-data/malformed.csv]: Now using only test files! committed here: 9c87ebd12f9eb4aef748f3416d2d9d2734d8e14e

**Pull request closed without merging**